### DISCUSSION TITLE
rabbitmq used disk space grows after upgrade to 3.13

### DISCUSSION THREAD

**Discussion by urusha** [rabbitmq/rabbitmq-server#10681] [Questions] [answered]
### Describe the bug

After upgrade from 3.12.13 to 3.13.0 used disk space grows up.

Average messages sum (count, bytes, publish, delivery) remained the same.
While `/metrics/per-object` shows `sum(rabbitmq_queue_messages_bytes)` about 35G, vhost directory on disk uses 1.5Tb:

```
# du -sh *
1.5T	msg_store_persistent
4.0K	msg_store_transient
30G	queues
20K	recovery.dets
# ls msg_store_persistent | wc -l
97035

```

`rabbitmq.conf ` contains `classic_queue.default_version = 2` (was after upgrading to 3.12)

After restart of rabbitmq folder size drops to 500G (still too much) and then begins to grow again.

### Reproduction steps

1. upgrade from 3.12 to 3.13
2. wait till directory size grows up

### Expected behavior

Directory size =~ total size of all queues.

### Additional context

![image](https://github.com/rabbitmq/rabbitmq-server/assets/1746504/d324d863-e6a8-4d0b-bb52-58069584d002)
rabbitmqctl version
3.13.0
erlang 26.2.2-1
debian 12 amd64

**Comment by lukebakken** [comment]
@urusha thanks for using RabbitMQ.

You have not provided enough information and are basically asking us to spend time guessing how to reproduce your issue.

* Please export your RabbitMQ definitions and _ATTACH_ the file.
* Attach your complete RabbitMQ configuration files.
* Describe your upgrade process in detail.
* What is your workload? Can you provide PerfTest arguments to mimic it? https://perftest.rabbitmq.com/

Ideally, you would provide a docker compose project that represents your workload and starts with RabbitMQ 3.12.

**Reply by urusha** [reply]
Sorry, misinformed you, OS is ubuntu 22.04 amd64.
Configuration files:
enabled_plugins
```
[rabbitmq_management,rabbitmq_prometheus,rabbitmq_shovel,rabbitmq_shovel_management].
```
rabbitmq.conf
```
classic_queue.default_version = 2
disk_free_limit.absolute = 10GB
prometheus.tcp.ip = 127.0.0.1
```
Upgrade process:
```
apt install rabbitmq-server
```
this restarted  rabbitmq-server systemd unit successfully, erlang was already the latest version.

Definitions (passwords,names modified):
[defs3.json](https://github.com/rabbitmq/rabbitmq-server/files/14501329/defs3.json)

Here is last week queue sizes in messages (not stacked) and bytes (stacked):
![image](https://github.com/rabbitmq/rabbitmq-server/assets/1746504/d819a88b-b9ec-4e47-bd28-a3ec9b65f2de)

I can provide other monitoring data if needed.

Will try to make perftests later.

**Reply by urusha** [reply]
just restarted - size dropped:
```
du -sh *
203G	msg_store_persistent
4.0K	msg_store_transient
24G	queues
20K	recovery.dets
```
but grows again.

**Reply by michaelklishin** [reply]
All that output tells us is that most of the data is used by classic queues. That's exactly why I have mentioned the `-d 3` part of `du -h -d 3`.

Your queues use 10 priorities each. Internally that means you have 10 classic queues with the same semantic name and a different consumption mechanism. 

Not only most systems [do not need even 10 priorities](https://www.rabbitmq.com/docs/priority#resource-usage),
I'm quite certain that CQv2 were tested with regular non-priority queues 99% of the time.

With peaks up to 70M+ messages, you will see periods of growing and shrinking disk space usage.

From there, besides reducing the number of priorities (which can only be changed by deleting and re-declaring the queues), I'd see what your applications do.

There may be genuine backlogs of messages, growing as you observe disk space usage. There may be an infinite or very high [prefetch level](https://www.rabbitmq.com/docs/confirms#channel-qos-prefetch) that means that a lot of messages will NOT be accounted in the "ready" state but would still consume disk space at the very least until they are confirmed. And yet they may or may not be accounted by the metric you expect actual disk space usage to follow closerly.

Finally, there is an extremely common recommendation of not sending around large (say, 10s of MBs) where possible, or at least use compression on the application side. A 100 MB file belongs to a blob store, with its ID passed around in messages.

**Reply by michaelklishin** [reply]
Moreover, both management UI and detailed Prometheus metrics should have enough to help correlate the number of Ready, Unacknowledged messages vs. computed disk footprint.

There may be a very small (even one) number of queues with a backlog of very large messages. When the node was restarted, there are two possible scenarios that can lead to a lot of data being removed:

1. All outstanding (unconfirmed) deliveries will be requeued
2. If applications publish messages [as transient](https://www.rabbitmq.com/docs/publishers#message-properties), those messages will be deleted on node startup/during queue data recovery

Of course, both of those factors can be in play at the same time, and they correspond reasonably well with the hypothesis that a small number of queues has a backlog of large messages that were published as transient. CQv2 will store such "transient" messages on disk just like any other but they would still be deleted on node restart,

Then after the node is restarted, the backlog begin accumulating again.

[Queue length limits](https://www.rabbitmq.com/docs/maxlength) and [message or even queue TTL](https://www.rabbitmq.com/docs/ttl) are two guardrail features available for such cases. They have been around forever. They are not commonly used with priority queues but both should do their job reasonably well, as long as every sub-queue is moving (not stuck waiting for a single consumer).

**Comment by michaelklishin** [comment]
@urusha I'm afraid the assumption that `rabbitmq_queue_messages_bytes` is all that node store on disk is a very naive one.

Quorum queues and streams can store a large amount of data on disk and there's nothing new in 3.13 compared to 3.12. Not only every message has protocol-level metadata, internal format metadata (which did change in 3.13, as release notes state) but that data is not deleted the moment a message is consumed.

Specifically, quorum queues store the entire Raft log, including messages and other state machine transitions, until the moment when the oldest **unacknowledged** message in the log is consumed and acknowledged. Stuck or slow consumers can affect this a great deal, which is why a [consumer delivery timeout](https://www.rabbitmq.com/docs/consumers#acknowledgement-timeout) exists: to make sure a stuck consumer does not prevent a quorum queue from reclaiming disk space.

**New** clusters can use a [different Raft WAL segment size](https://www.rabbitmq.com/docs/quorum-queues#resource-use). Changing this on existing clusters is dangerous and must not be attempted.

Streams retain as much data as you configure.

**Comment by michaelklishin** [comment]
The seesaw pattern on the chart above is large what I'd expect from quorum queues with many workloads, and streams as well. You have peaks and you have troughs. Few workloads result in a very even disk space use over time.

Without inspecting node data directory using `du -h -d 3 /path/to/node/data/directory` or something similar, we can only guess what specifically takes the space but usually the key contributor is the Raft log of a set of quorum queues.

CQv2 may result in higher peaks but we do not see this in practice, or somehow I've missed those cases. @lhoguin and @mkuratczyk would know better than I do.

**Reply by urusha** [reply]
@michaelklishin We only use classic queues with one-and-the-only node.  upgrade from 3.12 was at 2024-02-29 11:17. Before that time we've never had such issues with disk space (including 3.12.x with classic queues v2, for weeks). Overall data throughput didn't grow up either. Space issues started exactly after upgrade. Anyway, metadata size ~50x of actual data size isn't normal.

**Reply by lukebakken** [reply]
We can't help you without more information. Please see [my comment](https://github.com/rabbitmq/rabbitmq-server/discussions/10681#discussioncomment-8683630)

**Comment by lukebakken** [comment]
@urusha to confirm -

* You **NEVER** saw this issue using RabbitMQ `3.12.x`, and you are certain that your workload was identical to what it is using `3.13.0`
* You have _always_ used the `classic_queue.default_version = 2` setting in your environments.

**Reply by urusha** [reply]
1. yes
2. yes since 3.12

**Comment by lhoguin** [comment]
Hello, since space is not being reclaimed it's possible there is an issue with GCing messages, which has changed in 3.13. Please set the log levels to `debug` https://www.rabbitmq.com/docs/logging#log-levels and then provide the log lines for the following messages (filtering on `reclaimed` should help):

* https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_msg_store.erl#L1799-L1800
* https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_msg_store.erl#L1906

**Reply by urusha** [reply]
Hello, here it is:
[reclaimed.log.gz](https://github.com/rabbitmq/rabbitmq-server/files/14511095/reclaimed.log.gz)

**Reply by lhoguin** [reply]
Thank you. It appears most files get compacted but not all eventually get deleted. This may be because the file still has messages referred to.

I wonder if the files do get truncated as well though, could you provide the same log but with the following lines also included:

* https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_msg_store.erl#L1874-L1875
* https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_msg_store.erl#L1885

Thank you.

**Reply by urusha** [reply]
`grep -ahE 'reclaimed|Deferring|Truncated' /var/log/rabbitmq/*.log`
[rab.log.gz](https://github.com/rabbitmq/rabbitmq-server/files/14514441/rab.log.gz)

**Reply by lhoguin** [reply]
Thank you. I think it's simply not able to deal with the amount of compaction it is trying to do, resulting in not enough files getting truncated or deleted to reclaim the space. Evidently you have a lot of files that should be deleted - they are when the node is starting, before the node starts to do too many things and it gets behind again.

To confirm this I would like you to run `rabbitmq-diagnostics observer`. If you then press `m + Enter` to sort by memory, or `r + Enter` to sort by reductions, you should see a process named `rabbit_msg_store_gc` on your screen. Take a screenshot of these screens. On the left of the line is a number. By pressing that number + Enter, you will get the details of the process. You can then press `S + Enter` to get the state. Make sure your window is large enough, then screenshot this screen.

If this verifies, a workaround would be to get faster hardware (CPU or disk depending on what the bottleneck is). But we do want to get this fixed. My guess is that we simply try to queue up too much compaction, whereas in 3.12 and earlier there was a limit in the number of compacted files queued up at once. I made the wrong assumption that hardware was always fast enough that it was no longer necessary. Thank you!

**Reply by lhoguin** [reply]
We can reproduce behavior similar to what you observe and are working on fixing that, but having confirmation would be great still.

**Reply by urusha** [reply]
![image](https://github.com/rabbitmq/rabbitmq-server/assets/1746504/7b59573a-1970-42ae-9724-5dcc170416df)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/1746504/0d238101-b232-4b5a-978a-6116552a9f43)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/1746504/39de7a94-8021-45dd-9546-3d228327c846)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/1746504/226013d5-46c4-4af5-a8e7-e20b14ad318a)

**Reply by lhoguin** [reply, marked as answer]
Thank you. This is consistent with what we are observing. I am working on a fix. For what it's worth I do not think it's related to hardware anymore, we have found a small piece of code that is very inefficient when there are many messages that make compactions much slower than they should be. So for the time being we will put back a limit on the number of compactions in-flight.

**Reply by urusha** [reply]
@lhoguin thanks, glad to help! Is there a patch with the fix (and how to apply it), or a fixed package? we have to restart rabbit every several hours - and since it doesn't die gracefully (killed by systemd timeout), on start we have to wait for another batch of time till all topics get reindexed.
```
Mar 11 03:59:07 rabbit01 systemd[1]: Stopping RabbitMQ broker...
Mar 11 03:59:08 rabbit01 rabbitmqctl[93491]: Shutting down RabbitMQ node rabbit@rabbit01 running at PID 78243
Mar 11 04:00:37 rabbit01 systemd[1]: rabbitmq-server.service: Stopping timed out. Terminating.
Mar 11 04:00:37 rabbit01 systemd[1]: rabbitmq-server.service: Control process exited, code=killed, status=15/TERM
Mar 11 04:00:38 rabbit01 rabbitmqctl[93491]: Error:
Mar 11 04:00:38 rabbit01 rabbitmqctl[93491]: nodedown
Mar 11 04:02:08 rabbit01 systemd[1]: rabbitmq-server.service: State 'stop-sigterm' timed out. Killing.

**Reply by lhoguin** [reply]
I am working on this branch: https://github.com/rabbitmq/rabbitmq-server/pull/10696 - the code is currently being tested but is showing good results so far.

Depending on what you are using there is an OCI image built automatically as part of the GH Actions. Needless to say this is not a build we provide support for.

I expect testing to be done today. A v3.13.1 containing the fix should release this week.

**Reply by lukebakken** [reply]
@urusha please see this page:

https://hub.docker.com/r/pivotalrabbitmq/rabbitmq/tags

```
docker pull pivotalrabbitmq/rabbitmq:loic-cq-defer-gc-active-files-otp-max-bazel
```

That will contain the code from this PR. If you can test it in your environment, that would be great. Hopefully you can use docker.

**Reply by michaelklishin** [reply]
To add to the above, if you need a different package type in order to test this, we can build you one. We would love to have some feedback of that kind before we ship `3.13.1`.

The original post says you run RabbitMQ on Debian. Does it mean you need a .deb package? We have a development versions repo on Cloudsmith.io, and it should be updated shortly after the backporting PR is merged.

**Reply by urusha** [reply]
docker is ok, thanks. Currently it's starting with lots of `Rebuilding message location index from ...`, `Deleted empty file`. I will let you know about the results.

**Reply by urusha** [reply]
It seems that used disk space is not growing like before, it's stable, but rabbit still uses much more space than before upgrade from 3.12. Is there a way to clean it somehow without loosing data? Or we just need to wait until old artifacts are cleaned?
![image](https://github.com/rabbitmq/rabbitmq-server/assets/1746504/a06fac4b-6110-42fe-9419-2be5664df468)

**Reply by urusha** [reply]
Memory usage is also stable now.
![image](https://github.com/rabbitmq/rabbitmq-server/assets/1746504/78b28ba7-2480-4152-910c-a051550ceda2)

**Reply by lhoguin** [reply]
Thanks for the confirmation!

Rabbit will not start compaction on the files following startup recovery. It will only do so when at least one message gets acked from that file. So as your queues get consumed the amount of data remaining should diminish. I would expect a slow reduction in the amount of data used over the next few hours (or more, depending on how slowly the queues get consumed).

If you have queues that are consumed very slowly, this may take a while. If the data in some of the queues is not important (for example it is likely outdated), you can purge that queue and free some space. If you have queues that hold messages that are rarely accessed, you could consume and republish these messages to a new queue (or recreate the same queue). There's a number of other things that can be done but the gist of it is at least one of the messages in the older files need to be acked for compaction to occur.

For now I would recommend waiting at least a few hours to confirm that disk usage does go down (even if slowly) and report back.

**Reply by urusha** [reply]
After a while all the extra space was cleaned. Thanks a lot. Is it safe to upgrade currently running docker version to 3.13.1 (after release)? UI shows `RabbitMQ 4.0.0` version.

**Reply by lhoguin** [reply]
Thank you for the confirmation!

I don't think we have merged much that would prevent going to 3.13.1 at the time the PR was merged. The UI is saying this just because the docker image was built from the `main` branch, but at the time (and even still now) they were not very different.

**Reply by urusha** [reply]
We just tried to upgrade to `3.13.6 (26.2.5.2-1)` and getting queue crashes from time to time (classic v2). The same was with previous try - 3.13.2 release. We've moved back to docker build that time. But now after runing `3.13.6` for several hours, `pivotalrabbitmq/rabbitmq:loic-cq-defer-gc-active-files-otp-max-bazel` is crashing on startup with our current data. So we have unstable rabbitmq again and can't move back with the current data. We also tired 4.0-beta3 and experienced even more queue crashes. Never had such errors before `3.13`.
```
[error] <0.337985.0> operation queue.declare caused a channel exception not_found: queue 'queue' in vhost 'vhost' has crashed and failed to restart
[error] <0.337990.0> operation queue.declare caused a channel exception not_found: queue 'queue' in vhost 'vhost' has crashed and failed to restart
[error] <0.343839.0> Restarting crashed queue 'queue' in vhost 'vhost'.
```
The trigger for update to` 3.16.6` was running out of space on the disk (thought it was this bug again), but after further investigation we found that additional space was a valid data in one of the queues. So we fixed the space problem. And we have `disk_free_limit.absolute = 10GB`, so we always had real available bytes on the disk.
Now we have:
`pivotalrabbitmq/rabbitmq:loic-cq-defer-gc-active-files-otp-max-bazel` crashing on startup.
`3.13.6` with crashing queues (sometimes) - running it
`4.0-beta3` with crashing queues (constantly)

All load/queues setup is pretty the same as before.
What additional info should I provide to you?
We would really like to stay on some stable version again :)

**Reply by michaelklishin** [reply]
A node that has run out of disk space is not guaranteed to recover. You never know what was not written to disk.

Such nodes must be recreated from scratch. It's been highly recommended for years.

Disk space is the cheapest resource of them all. Overprovision it and use queue length limits. Again, both ideas/features have been around for years.

**Reply by michaelklishin** [reply]
Having "real available bytes on disk" should not be celebrated as an accomplishment, it is the most basic requirement of any deployment, and it's always been the case. 

The [Production Checklist doc guide](https://www.rabbitmq.com/docs/production-checklist#storage) explicitly says that free disk space should be overprovisioned w.r.t. the workload.

"Running out of disk space" is not necessarily a bug, in fact, in 99% of cases it is an indication of the lack of adequate queue length limits, lack of consumers online, stuck consumers and so on.

**Comment by michaelklishin** [comment]
The original issue reported here was resolved, and I do not recall it being reported elsewhere since March. 

@urusha please use the recommendations above (from today) and start a new one if you have something else to report, with a detailed set of steps to reproduce. Most likely we will need a data directory from the node that fails to start.

**Comment by michaelklishin** [comment]
```
The trigger for update to 3.16.6 was running out of space on the disk (thought it was this bug again), 
but after further investigation we found that additional space was a valid data in one of the queues. 

So we fixed the space problem. And we have disk_free_limit.absolute = 10GB,
so we always had real available bytes on the disk.

Now we have:
pivotalrabbitmq/rabbitmq:loic-cq-defer-gc-active-files-otp-max-bazel crashing on startup.
```

Like I said above, a node that has run out of disk space won't always be able to recover. Overprovision free disk space, put adequate guardrails such as queue length limits and max message size in place.

The issue with CQ compaction falling behind, reported in this thread in March, has been addressed and never reported again.

A recovering node cannot know what **was not** written to disk, so there always will be scenarios where it won't be able to safely recover. If your node has run out of disk space, consider it to be ready to be replaced.
